src/linux/sections.rs

```diff
@@ -14,159 +14,239 @@ use crate::{
     },
     minidump_format::*,
 };
-use std::convert::TryInto;
-use std::io::Write;
+use scroll::ctx::{SizeWith, TryIntoCtx};
 
-pub type Buffer = std::io::Cursor<Vec<u8>>;
 type WriteResult<T> = std::result::Result<T, MemoryWriterError>;
 
+macro_rules! size {
+    ($t:ty) => {
+        <$t>::size_with(&scroll::Endian::Little)
+    };
+}
+
+pub struct Buffer {
+    inner: Vec<u8>,
+}
+
+impl Buffer {
+    pub fn with_capacity(cap: usize) -> Self {
+        Self {
+            inner: Vec::with_capacity(cap),
+        }
+    }
+
+    #[inline]
+    pub fn position(&self) -> u64 {
+        self.inner.len() as u64
+    }
+
+    #[inline]
+    #[must_use]
+    fn reserve(&mut self, len: usize) -> usize {
+        let mark = self.inner.len();
+        self.inner.resize(self.inner.len() + len, 0);
+        mark
+    }
+
+    #[inline]
+    fn write<N, E>(&mut self, val: N) -> Result<usize, E>
+    where
+        N: TryIntoCtx<scroll::Endian, Error = E> + SizeWith<scroll::Endian>,
+        E: From<scroll::Error>,
+    {
+        self.write_at(self.inner.len(), val)
+    }
+
+    fn write_at<N, E>(&mut self, offset: usize, val: N) -> Result<usize, E>
+    where
+        N: TryIntoCtx<scroll::Endian, Error = E> + SizeWith<scroll::Endian>,
+        E: From<scroll::Error>,
+    {
+        let to_write = size!(N);
+        let remainder = self.inner.len() - offset;
+        if remainder < to_write {
+            self.inner
+                .resize(self.inner.len() + to_write - remainder, 0);
```

Comment:
If we are out of memory, this will result in a panic, right? With the old Cursor, I _think_ we got an actual error, which could have been handed out, but I'm actually not sure. Maybe Cursor.write() would also panic in case of OOM.

Reply:
Nope, Vec's `Write` implementation doesn't handle OOM either https://doc.rust-lang.org/src/std/io/impls.rs.html#333-338